File: code-tiny/tools/spring/adapters.py
from __future__ import annotations

import os
import re
from typing import Iterable, List, Protocol, Sequence

from tools.spring.detector import read_limited
from tools.spring.models import Diagnostic, LanguageSourceFact


_PACKAGE_RE = re.compile(r"^\s*package\s+([A-Za-z_][\w.]*)(?:\s*;)?", re.MULTILINE)
_ANNOTATION_RE = re.compile(r"@([A-Za-z_][\w.]*)")
_JAVA_DECL_RE = re.compile(r"\b(?:class|interface|enum|record)\s+([A-Za-z_]\w*)")
_KOTLIN_DECL_RE = re.compile(r"\b(?:class|interface|object|data\s+class|enum\s+class)\s+([A-Za-z_]\w*)")


class LanguageAdapter(Protocol):
    language: str
    extensions: Sequence[str]

    def collect(self, root: str, rel_paths: Iterable[str]) -> List[LanguageSourceFact]:
        ...


class _BaseAdapter:
    language = ""
    extensions: Sequence[str] = ()
    declaration_pattern: re.Pattern[str]

    def collect(self, root: str, rel_paths: Iterable[str]) -> List[LanguageSourceFact]:
        facts: List[LanguageSourceFact] = []
        for rel_path in sorted(set(rel_paths)):
            lower = rel_path.lower()
            if not lower.endswith(tuple(self.extensions)):
                continue
            text = read_limited(os.path.join(root, rel_path), limit=512 * 1024)
            package = _extract_package(text)
            declarations = tuple(self.declaration_pattern.findall(text))
            annotations = tuple(sorted(set(_ANNOTATION_RE.findall(text))))
            facts.append(
                LanguageSourceFact(
                    language=self.language,
                    file_path=rel_path,
                    source_symbol_id=f"{self.language}::file::{rel_path}",
                    package_name=package,
                    declarations=declarations,
                    annotations=annotations,
                    parser_status="base_adapter_pending",
                )
            )
        return facts


class JavaLanguageAdapter(_BaseAdapter):
    language = "java"
    extensions = (".java",)
    declaration_pattern = _JAVA_DECL_RE


class KotlinLanguageAdapter(_BaseAdapter):
    language = "kotlin"
    extensions = (".kt", ".kts")
    declaration_pattern = _KOTLIN_DECL_RE


def _extract_package(text: str) -> str:
    match = _PACKAGE_RE.search(text)
    return match.group(1) if match else ""


def adapters_for_languages(languages: Sequence[str]) -> List[LanguageAdapter]:
    requested = set(languages)
    adapters: List[LanguageAdapter] = []
    if "java" in requested:
        adapters.append(JavaLanguageAdapter())
    if "kotlin" in requested:
        adapters.append(KotlinLanguageAdapter())
    return adapters

File: rust/crates/analyzer-jvm-overlays/src/spring/adapters.rs
//! Port `tools/spring/adapters.py` — Java/Kotlin language source facts.

use regex::Regex;
use super::models::LanguageSourceFact;
use crate::pyutil::read_limited;
use std::path::Path;

fn package_re() -> &'static Regex {
    static RE: std::sync::OnceLock<Regex> = std::sync::OnceLock::new();
    RE.get_or_init(|| Regex::new(r"(?m)^\s*package\s+([A-Za-z_][\w.]*)(?:\s*;)?").unwrap())
}

fn java_decl_re() -> &'static Regex {
    static RE: std::sync::OnceLock<Regex> = std::sync::OnceLock::new();
    RE.get_or_init(|| Regex::new(r"\b(?:class|interface|enum|record)\s+([A-Za-z_]\w*)").unwrap())
}

fn kotlin_decl_re() -> &'static Regex {
    static RE: std::sync::OnceLock<Regex> = std::sync::OnceLock::new();
    RE.get_or_init(|| Regex::new(r"\b(?:class|interface|object|data\s+class|enum\s+class)\s+([A-Za-z_]\w*)").unwrap())
}

fn annotation_re() -> &'static Regex {
    static RE: std::sync::OnceLock<Regex> = std::sync::OnceLock::new();
    RE.get_or_init(|| Regex::new(r"@([A-Za-z_][\w.]*)").unwrap())
}

struct LanguageAdapter {
    language: &'static str,
    extensions: &'static [&'static str],
    declaration_pattern: &'static Regex,
}

fn adapters_for_languages(languages: &[&str]) -> Vec<LanguageAdapter> {
    let mut adapters: Vec<LanguageAdapter> = Vec::new();
    if languages.contains(&"java") {
        adapters.push(LanguageAdapter {
            language: "java",
            extensions: &[".java"],
            declaration_pattern: java_decl_re(),
        });
    }
    if languages.contains(&"kotlin") {
        adapters.push(LanguageAdapter {
            language: "kotlin",
            extensions: &[".kt", ".kts"],
            declaration_pattern: kotlin_decl_re(),
        });
    }
    adapters
}

pub fn collect_language_facts(root: &Path, languages: &[&str], rel_paths: &[String]) -> Vec<LanguageSourceFact> {
    let mut facts: Vec<LanguageSourceFact> = Vec::new();
    for adapter in adapters_for_languages(languages) {
        let mut sorted: Vec<&String> = rel_paths.iter().collect();
        sorted.sort();
        sorted.dedup();
        for rel_path in sorted {
            let lower = rel_path.to_lowercase();
            if !adapter.extensions.iter().any(|ext| lower.ends_with(ext)) {
                continue;
            }
            let text = read_limited(&root.join(rel_path), 512 * 1024);
            let package = package_re()
                .captures(&text)
                .and_then(|c| c.get(1))
                .map(|m| m.as_str().to_string())
                .unwrap_or_default();
            let declarations: Vec<String> = adapter
                .declaration_pattern
                .captures_iter(&text)
                .filter_map(|c| c.get(1))
                .map(|m| m.as_str().to_string())
                .collect();
            let mut annotations: Vec<String> = annotation_re()
                .captures_iter(&text)
                .filter_map(|c| c.get(1))
                .map(|m| m.as_str().to_string())
                .collect();
            annotations.sort();
            annotations.dedup();
            facts.push(LanguageSourceFact {
                language: adapter.language.to_string(),
                file_path: rel_path.clone(),
                source_symbol_id: format!("{}::file::{}", adapter.language, rel_path),
                package_name: package,
                declarations,
                annotations,
                parser_status: "base_adapter_pending".to_string(),
            });
        }
    }
    facts
}
